target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,750 @@
+{
+	unsigned long irqflags;
+
+	spin_lock_irqsave(&rtcl_ccu->lock, irqflags);
+	rtcl_838x_sram_set_rate(clk_idx, reg->ctrl0, reg->ctrl1);
+	spin_unlock_irqrestore(&rtcl_ccu->lock, irqflags);
+	return 0;
+}
+
+static int rtcl_839x_set_rate(int clk_idx, const struct rtcl_reg_set *reg)
+{
+	unsigned long irqflags;
+	unsigned int mtflags, vpflags;
+
+	spin_lock_irqsave(&rtcl_ccu->lock, irqflags);
+	mtflags = dmt();
+	vpflags = dvpe();
+	rtcl_839x_sram_set_rate(clk_idx, reg->ctrl0, reg->ctrl1);
+	evpe(vpflags);
+	emt(mtflags);
+	spin_unlock_irqrestore(&rtcl_ccu->lock, irqflags);
+	return 0;
+}
+
+static int rtcl_set_rate(struct clk_hw *hw, unsigned long rate, unsigned long parent_rate)
+{
+	int tab_idx;
+	struct rtcl_clk *clk = rtcl_hw_to_clk(hw);
+	const struct rtcl_rtab_set *rtab = &rtcl_rtab_set[rtcl_ccu->soc][clk->idx];
+
+	if ((parent_rate != OSC_RATE) || (!rtcl_ccu->sram.vbase))
+		return -EINVAL;
+/*
+ * Currently we do not know if SRAM is stable on these devices. Maybe someone changes memory in
+ * this region and does not care about proper allocation. So check if something might go wrong.
+ */
+	if (*rtcl_ccu->sram.pmark != RTL_SRAM_MARKER) {
+		dev_err(&rtcl_ccu->pdev->dev, "SRAM code lost\n");
+		return -EINVAL;
+	}
+
+	tab_idx = (rate - rtcl_round_set[rtcl_ccu->soc][clk->idx].min)
+		/ rtcl_round_set[rtcl_ccu->soc][clk->idx].step;
+
+	if ((tab_idx < 0) || (tab_idx >= rtab->count) || (rtab->rset[tab_idx].rate != rate))
+		return -EINVAL;
+
+	switch (rtcl_ccu->soc) {
+	case SOC_RTL838X:
+		return rtcl_838x_set_rate(clk->idx, &rtab->rset[tab_idx]);
+	case SOC_RTL839X:
+		return rtcl_839x_set_rate(clk->idx, &rtab->rset[tab_idx]);
+	}
+	return -ENXIO;
+}
+
+static long rtcl_round_rate(struct clk_hw *hw, unsigned long rate, unsigned long *parent_rate)
+{
+	struct rtcl_clk *clk = rtcl_hw_to_clk(hw);
+	unsigned long rrate = max(clk->min, min(clk->max, rate));
+	const struct rtcl_round_set *round = &rtcl_round_set[rtcl_ccu->soc][clk->idx];
+
+	rrate = ((rrate + (round->step >> 2)) / round->step) * round->step;
+	rrate -= (rrate > clk->max) ? round->step : 0;
+	rrate += (rrate < clk->min) ? round->step : 0;
+	return rrate;
+}
+
+/*
+ * Initialization functions to register the CCU and its clocks
+ */
+
+#define SRAM_FUNC(SOC, PBASE, FN) ({					\
+	rtcl_##SOC##_sram_##FN = ((void *)&rtcl_##SOC##_dram_##FN	\
+				- (void *)&rtcl_##SOC##_dram_start)	\
+				+ (void *)PBASE; })
+
+static const struct clk_ops rtcl_clk_ops = {
+	.round_rate = rtcl_round_rate,
+	.recalc_rate = rtcl_recalc_rate,
+	.set_rate = rtcl_set_rate,
+};
+
+static int rtcl_ccu_create(struct device_node *np)
+{
+	int soc;
+
+	if (of_device_is_compatible(np, COMPAT_RTL838X))
+		soc = SOC_RTL838X;
+	else if (of_device_is_compatible(np, COMPAT_RTL839X))
+		soc = SOC_RTL839X;
+	else
+		return -ENXIO;
+
+	rtcl_ccu = kzalloc(sizeof(*rtcl_ccu), GFP_KERNEL);
+	if (IS_ERR(rtcl_ccu))
+		return -ENOMEM;
+
+	rtcl_ccu->soc = soc;
+	rtcl_ccu->dram.type = RTL_MC_MCR_DRAMTYPE(read_soc(RTL_MC_MCR));
```

Comment:
so this is a bit weird, because normally as a platform driver, you could call one of the of_ functions, that will give you the address you need, as per devicetree definition. I haven't seen where you call rtl_ccu_create, and this might be a mips/clock thing, but usually 'this is not the way'. I think ralink was also mips wasn't it? While they have some nice regmap driver, they also get the address in a sense from the dt (https://elixir.bootlin.com/linux/latest/source/drivers/clk/ralink/clk-mt7621.c#L492) (i lack proper examples that are not arm :p)

Reply:
I will keep my patch simple and will not fix missing memory controller DT registers.